### PR TITLE
Add busybox sh support

### LINKED ISSUE #2865: Improved busybox sh support
#### For new checks and feature suggestions
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

Busybox SH is an ash derivative with a few features from ksh.
Running some tests, I noticed the following SCxxxx errors do not apply to busybox sh

- SC3010: [[ ]] in place of [ ]
- SC3014:  == in place of =
- SC3020: &> 
- SC3046/SC3051: 'source' in place of '.'
- SC3048: prefixing signal names with 'SIG'
- SC3057: string indexing 
- SC3060: string replacement

```sh
#!/bin/busybox sh

 X="thisisastring"

stop() {
    echo stopping
    exit 0
}
# SC3048
trap stop EXIT SIGTERM SIGINT

# SC3010 and SC3014
[[ $X == "thisisastring" ]] && echo "this shell supports == and [[ ]]"

# SC3020
echo "busybox sh supports &>" &> /dev/null

# SC3046/SC3051
source /dev/null

# SC3057
echo "${X:0:4} shell supports ${X:7} indexing" # "this shell supports string indexing"

# SC3060
echo "${X/string/magic}"

# for SC3048
sleep 3600
```

and the output is:
```
$ ./busyboxsh.sh
this shell supports == and [[ ]]
this shell supports string indexing
thisisamagic
^Cstopping
stopping
```

shellcheck 
```
In busyboxsh.sh line 1:
#!/bin/busybox sh
^---------------^ SC2187 (warning): Ash scripts will be checked as Dash. Add '# shellcheck shell=dash' to silence.


In busyboxsh.sh line 10:
trap stop EXIT SIGTERM SIGINT
               ^-----^ SC3048 (error): In dash, prefixing signal names with 'SIG' is not supported.
                       ^----^ SC3048 (error): In dash, prefixing signal names with 'SIG' is not supported.


In busyboxsh.sh line 13:
[[ $X == "thisisastring" ]] && echo "this shell supports == and [[ ]]"
^-------------------------^ SC3010 (error): In dash, [[ ]] is not supported.


In busyboxsh.sh line 16:
echo "busybox sh supports &>" &> /dev/null
                              ^----------^ SC3020 (error): In dash, &> is not supported.


In busyboxsh.sh line 19:
source /dev/null
^--------------^ SC3046 (error): In dash, 'source' in place of '.' is not supported.
^--------------^ SC3051 (error): In dash, 'source' in place of '.' is not supported.


In busyboxsh.sh line 22:
echo "${X:0:4} shell supports ${X:7} indexing" # "this shell supports string indexing"
      ^------^ SC3057 (error): In dash, string indexing is not supported.
                              ^----^ SC3057 (error): In dash, string indexing is not supported.


In busyboxsh.sh line 25:
echo "${X/string/magic}"
      ^---------------^ SC3060 (error): In dash, string replacement is not supported.
```

## Improvement

Even if busybox sh will not be a first-class shell, would it be possible to silence those warnings/errors for busybox sh?

### PR DESCRIPTION
**Pull request by grische**
This adds support for Busybox' variant of Ash. 
Added tests for all manually verified options on BusyBox v1.30.1.

Fixes #2865

### CODE REVIEW

File: src/ShellCheck/Interface.hs [hunk 1]
```diff
@@ -221,7 +221,7 @@ newCheckDescription = CheckDescription {
     }
 
 -- Supporting data types
-data Shell = Ksh | Sh | Bash | Dash deriving (Show, Eq)
+data Shell = Ksh | Sh | Bash | Dash | Busyboxsh deriving (Show, Eq)
```
Comment (koalaman) [inline src/ShellCheck/Interface.hs:224]: `BusyboxSh` would be easier to read.

Any thoughts on `BusyboxSh` vs `BusyboxAsh`? I have no strong opinions either way.

File: src/ShellCheck/Data.hs [hunk 2]
```diff
@@ -156,6 +156,8 @@ shellForExecutable name =
         "sh"    -> return Sh
         "bash"  -> return Bash
         "bats"  -> return Bash
+        "busybox sh"  -> return Busyboxsh
+        "busybox ash"  -> return Busyboxsh
         "dash"  -> return Dash
         "ash"   -> return Dash -- There's also a warning for this.
```
Comment (koalaman) [inline src/ShellCheck/Data.hs:162]: Any thoughts on whether Dash or BusyboxSh is a better match for `ash`?

**Review by koalaman — COMMENTED** [review]
Nice, looks pretty much good to go!

File: src/ShellCheck/Interface.hs (hunk 1 above)
Comment (grische) [reply src/ShellCheck/Interface.hs:224]: I went for busybox sh, as it seems to strongly diverge from other Ash implementations.

File: src/ShellCheck/Data.hs (hunk 2 above)
Comment (grische) [reply src/ShellCheck/Data.hs:162]: Dash is probably the more canonical Ash. I wouldn't change that.

File: src/ShellCheck/Interface.hs (hunk 1 above)
Comment (grische) [reply src/ShellCheck/Interface.hs]: renamed it to `BusyboxSh`

**Comment by grische** [comment]
@koalaman is anything keeping this PR from being merged?

**Pull request merged**